### PR TITLE
designer improvements

### PR DESCRIPTION
**Pull request by donandren**
- attempt to fix lock of build output assemblies (used for preview and IntelliSense), by copying them to temp folder
- added option to disable preview, so only IntelliSense is active
- updated IntelliSense  system to be better supporting avalonia 0.9 features like avares compiled resources

available for early feedback: https://drive.google.com/open?id=1IVyIcFTUA17LMZ_QFr-_JCGepgyG3pwb

- partially solves the issue #104 (Enable/Disable should restart designer)
- #98 Fixed
- #60 Fixed
- #54 Fixed

### CODE REVIEW

**Comment by grokys** [comment]
Thanks @donandren - do you think it'd be possible to split this up into separate PRs for easy review? In particular it would be nice to have the "copy to temp folder" feature as a separate PR so that we can discuss it separately, but also it makes it easier to review.

**Comment by donandren** [comment]
The PR is more or less complete and tested, the extension is working decent, stable and with major improvements after this pr changes, it wouldn't make sense for me to put some additional work to split only the copy to temp functionality (it's really a small part of the PR a commit ), as it might take more time to do that instead fix or improve something else in my free time for avalonia (extracting a commit(s) then later dealing with conflicts can be a time eater :( don't want to risk broke what is already working pretty good)

**Comment by grokys** [comment]
Ok, no problem - thanks for the work you've done on it!

File: .gitmodules [hunk 1]
```diff
@@ -3,4 +3,4 @@
 	url = https://github.com/AvaloniaUI/Avalonia
 [submodule "submodules/Avalonia.Ide"]
 	path = submodules/Avalonia.Ide
-	url = https://github.com/grokys/Avalonia.Ide
+	url = https://github.com/donandren/Avalonia.Ide
```
Comment (danwalmsley) [inline .gitmodules]: We need to accept your changes into Avalonia.Ide

Comment (danwalmsley) [reply .gitmodules]: prob should be: https://github.com/kekekeks/Avalonia.ide

@grokys @kekekeks maybe we need to get his repo moved to Avalonia organization?

Comment (donandren) [reply .gitmodules]: that's correct. there is still outstanding pr  https://github.com/kekekeks/Avalonia.Ide/pull/22 which is not merged, then submodule can be forwarded to the original

File: AvaloniaVS/Views/AvaloniaDesigner.xaml.cs [hunk 2]
```diff
@@ -185,11 +222,19 @@ public void Dispose()
...
+            var assemblyPath = SelectedTarget?.XamlAssembly;
+            var executablePath = SelectedTarget?.ExecutableAssembly;
+
             Process.FrameReceived -= FrameReceived;
 
             _throttle.Dispose();
             previewer.Dispose();
             Process.Dispose();
+
+            if (!alreadyDisposed && assemblyPath != null && executablePath != null)
```
Comment (kekekeks) [inline AvaloniaVS/Views/AvaloniaDesigner.xaml.cs]: This doesn't seem reliable

File: AvaloniaVS/Views/AvaloniaDesigner.xaml.cs [hunk 3]
```diff
@@ -284,56 +331,156 @@ where IsValidOutput(output)
...
+            var executableDirDesigner = Path.Combine(designerTempDir, "exe");
+            var assemblyDirDesigner = Path.Combine(designerTempDir, "asm");
 
-            var assemblyPath = SelectedTarget?.XamlAssembly;
-            var executablePath = SelectedTarget?.ExecutableAssembly;
+            return (executableDirDesigner, assemblyDirDesigner);
+        }
 
-            if (assemblyPath != null && executablePath != null)
+        private (string executablePath, string assemblyPath) TryPrepareDesignTempData(string executablePath, string assemblyPath)
```
Comment (kekekeks) [inline AvaloniaVS/Views/AvaloniaDesigner.xaml.cs]: This code doesn't clean the previous data from "shadow copy" directory after re-build which potentially breaks user's app

File: AvaloniaVS/Views/AvaloniaDesigner.xaml.cs [hunk 4]
```diff
@@ -284,56 +331,156 @@ where IsValidOutput(output)
             Log.Logger.Verbose("Finished AvaloniaDesigner.LoadTargetsAsync()");
         }
 
-        private async Task StartProcessAsync()
+        static private (string executableDir, string assemblyDir) GetDesignTempDirs(string executablePath, string assemblyPath)
         {
-            Log.Logger.Verbose("Started AvaloniaDesigner.StartProcessAsync()");
+            //let's try use very short folder names as limit for directory path is 248 chars
+            var designerTempDir = Path.Combine(Directory.GetParent(Path.GetDirectoryName(executablePath)).FullName, "dttmp");
```
Comment (kekekeks) [inline AvaloniaVS/Views/AvaloniaDesigner.xaml.cs]: This should at least be configurable by MSBuild properties.

One of the problems with shadow copying assemblies is that previewer effectively *runs the entire user's app*, which might need some data from other solution folders. The user code might have trouble finding those if we copy files to some random location.

Another problem is that people sometimes change their project output directory or just expect it not to have some random files laying around.

Comment (donandren) [reply AvaloniaVS/Views/AvaloniaDesigner.xaml.cs]: I can't see a potential problem here, as designer temp folder logic is chosen with that in mind.
if output folder of an app is:
`....\ApplicationOutput\bin\Debug\netcoreapp2.0` 
then designer temp folder will be:
`....\ApplicationOutput\bin\Debug\dttmp` 
so all the relative and absolute paths from the app if any dependencies should work fine.
Agree user might not like copy files to a random location, but not sure we can fix build output locks with different approach than copy output to a temp folder :(
we can make attempt to have settings in designer or ms build property where and what should be the name of the temp folder, but doubt anybody will use that option.

File: AvaloniaVS/Views/AvaloniaDesigner.xaml.cs (hunk 3 above)
Comment (donandren) [reply AvaloniaVS/Views/AvaloniaDesigner.xaml.cs]: binaries in output folder are not deleted on purpose, they are overridden with new files only if changed, so  we don't lock ourselves while lunching the designer process. Even if something wrong happend (unpredictable), shouldn't be a big issue as designer will clean up the temp folder when closed

File: AvaloniaVS/Views/AvaloniaDesigner.xaml.cs (hunk 2 above)
Comment (donandren) [reply AvaloniaVS/Views/AvaloniaDesigner.xaml.cs]: dispose might be called twice, but haven't been able to understand why

File: AvaloniaVS/Views/AvaloniaDesigner.xaml.cs (hunk 3 above)
Comment (kekekeks) [reply AvaloniaVS/Views/AvaloniaDesigner.xaml.cs]: There might be files that shouldn't exist anymore after rebuild.

File: .gitmodules (hunk 1 above)
Comment (grokys) [reply .gitmodules]: You shouldn't need to change the url of the submodule, the commit will be available from the original URL even if it was made in a fork.

**Comment by grokys** [comment]
When I get time later this week, I'll try to split this PR up into its component parts. @donandren in future if you could open separate PRs for each feature/bugfix it would be really helpful:

1. It makes it easier to review; otherwise:
  - It's hard to know what change relates to which fix/feature
  - One feature/fix can hold up the other featrues/fixes in the PR (such as "copy output directory")
2. It makes it easier to understand why a change was made in future (I've been needing this a lot recently!)
3. It makes it easier to revert a change should it cause problems

File: AvaloniaVS/Services/IAvaloniaVSSettings.cs [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+using System.ComponentModel;
+using Microsoft.VisualStudio.ComponentModelHost;
+using Serilog.Events;
+
+namespace AvaloniaVS.Services
+{
+    public enum DesignerViewType
+    {
+        Split,
+        Design,
```
Comment (grokys) [inline AvaloniaVS/Services/IAvaloniaVSSettings.cs]: What is the use-case for just displaying the preview without the XAML?

Comment (grokys) [reply AvaloniaVS/Services/IAvaloniaVSSettings.cs]: Ahh, never mind - it's the default view, it can still be switched ;)

**Comment by grokys** [comment]
Unchecking "Preview Enabled" seems to cause endless exceptions to be thrown:

image

File: AvaloniaVS/AvaloniaPackage.cs [hunk 6]
```diff
@@ -15,10 +17,12 @@ namespace AvaloniaVS
     [InstalledProductRegistration("#110", "#112", "1.0", IconResourceID = 400)]
     [PackageRegistration(UseManagedResourcesOnly = true, AllowsBackgroundLoading = true)]
     [ProvideAutoLoad(VSConstants.UICONTEXT.SolutionExistsAndFullyLoaded_string, PackageAutoLoadFlags.BackgroundLoad)]
-    [ProvideEditorExtension(typeof(EditorFactory), ".paml", 100, NameResourceID = 113, DefaultName = "Avalonia Xaml Editor")]
+    [ProvideEditorExtension(typeof(EditorFactory), ".paml", 100, NameResourceID = 113, EditorFactoryNotify = true, ProjectGuid = VSConstants.UICONTEXT.CSharpProject_string, DefaultName = Name)]
+    [ProvideEditorExtension(typeof(EditorFactory), ".xaml", 0x40, NameResourceID = 113, EditorFactoryNotify = true, ProjectGuid = VSConstants.UICONTEXT.CSharpProject_string, DefaultName = Name)]
```
Comment (grokys) [inline AvaloniaVS/AvaloniaPackage.cs]: Why were these two changes necessary? What issue/feature/bugfix do they relate to?

File: AvaloniaVS/Views/AvaloniaDesigner.xaml [hunk 7]
```diff
@@ -11,55 +11,75 @@
...
 
         <DockPanel Name="toolbar"
                    Grid.Row="0"
                    Margin="4"
                    VerticalAlignment="Top">
-            <ComboBox DockPanel.Dock="Right"
+            <StackPanel Orientation="Horizontal" DockPanel.Dock="Right">
+                <CheckBox Content="Preview Enabled" IsChecked="{Binding IsPreviewEnabled, ElementName=root, Mode=TwoWay}" Margin="0 0 10 0" 
+                          VerticalAlignment="Center" local:VsTheme.UseVsTheme="True"/>
+                <ListBox Margin="0 0 10 0" BorderThickness="0"
```
Comment (grokys) [inline AvaloniaVS/Views/AvaloniaDesigner.xaml]: I wonder if we need to put this in the options dialog? My initial thought is that the designer should just remember the last mode used when opening an new designer.

File: AvaloniaVS/AvaloniaPackage.cs (hunk 6 above)
Comment (grokys) [reply AvaloniaVS/AvaloniaPackage.cs]: The commit says "register extension in open with context menu" but I'm not sure what that means...

File: AvaloniaVS/IntelliSense/XamlCompletionCommandHandler.cs [hunk 8]
```diff
@@ -127,51 +127,58 @@ private bool HandleSessionUpdate(char c)
...
+            }
+
+            var selected = _session.SelectedCompletionSet.SelectionStatus.Completion as XamlCompletion;
+
+            if (!_session.SelectedCompletionSet.SelectionStatus.IsSelected || selected == null)
+            {
+                return false;
+            }
+
+            if (c == '/')
```
Comment (grokys) [inline AvaloniaVS/IntelliSense/XamlCompletionCommandHandler.cs]: What issue is this fixing? It seems to be making sure the element is closed when `/` is typed, but as far as I can tell that already worked.

File: AvaloniaVS/AvaloniaPackage.cs (hunk 6 above)
Comment (donandren) [reply AvaloniaVS/AvaloniaPackage.cs]: image
option in dialog after `Open With...`

Comment (donandren) [reply AvaloniaVS/AvaloniaPackage.cs]: that's really useful for xaml files with non avalonia namespace base class

File: AvaloniaVS/Views/AvaloniaDesigner.xaml (hunk 7 above)
Comment (donandren) [reply AvaloniaVS/Views/AvaloniaDesigner.xaml]: it's different, we have settings which are persisted (in Settings) and one for per editor. 
my workflow is like that:
- disabled preview and using only code completion
- enable preview manually to check something in ui, then switch back to xaml only edit (don't want persisted in another editor instances/xaml files)

File: AvaloniaVS/IntelliSense/XamlCompletionCommandHandler.cs (hunk 8 above)
Comment (donandren) [reply AvaloniaVS/IntelliSense/XamlCompletionCommandHandler.cs]: that's a code which i don't have a clue , but didn't wanted to delete it as i don't understand it.
But tried to simplify the logic here and extract the unknown part as it was having wierd logic that was breaking completion of `avares://Assembly/Resource.xaml` like urls

File: AvaloniaVS/Services/IAvaloniaVSSettings.cs (hunk 5 above)
Comment (donandren) [reply AvaloniaVS/Services/IAvaloniaVSSettings.cs]: designer is also xaml editor with completions, there is a use case to use it only for xaml code editing and no preview

File: .gitmodules (hunk 1 above)
Comment (donandren) [reply .gitmodules]: ha i didn't knew that

**Comment by grokys** [comment]
Regarding intellisense, I see it working when filling in `x:Class` attributes, but I can't get it to work when doing e.g. `<Image Source="` - should it be popping up intellisense there?

**Comment by grokys** [comment]
What is the use case for having both a "Preview Enabled" checkbox and a "Design View" option? When would you want to disable the previewer while viewing the design view and when would you want the previewer enabled when in the source view?

To my mind, the previewer should be started/stopped based on which view you're currently in. If the problem is #104 then IMO a button should appear when the previewer process stops to allow you to restart it.

**Comment by donandren** [comment]
> What is the use case for having both a "Preview Enabled" checkbox and a "Design View" option? When would you want to disable the previewer while viewing the design view and when would you want the previewer enabled when in the source view?
> 
> To my mind, the previewer should be started/stopped based on which view you're currently in. If the problem is #104 then IMO a button should appear when the previewer process stops to allow you to restart it.

i think we can live without enable preview checkbox, when we have split/code/preview options, although it might be useful is some scenarios, like disable preview  until you type xaml without changing layout then enable it back for reference, wouldn't mind we drop it tho

**Comment by donandren** [comment]
> Regarding intellisense, I see it working when filling in `x:Class` attributes, but I can't get it to work when doing e.g. `<Image Source="` - should it be popping up intellisense there?

for `<Image Source="`  onyl resm like resources are completed
`avares://` completion is working for xaml avalonia resources without x:Class only - for sure there is a room for improvements

**Comment by donandren** [comment]
@grokys  i can start working on this pr notes after your pr's are merged

**Comment by grokys** [comment]
@donandren all PRs merged. I'll close this PR now as there are a lot of conflicts.

**Comment by donandren** [comment]
@grokys  a couple of major fixes hasn't been missed from this pr in the small pr's that has been made. I'll keep this open until i verify all important pieces are translated to another pr's

**Comment by donandren** [comment]
> @donandren all PRs merged. I'll close this PR now as there are a lot of conflicts.

this was expected

**Pull request closed without merging**